File: Java/ScientificCalculator/src/functions/Factorial.java
package functions;

public class Factorial {

    // ==========================
    // Factorial
    // ==========================

    public static long factorial(int n) {

        if (n < 0) {
            throw new ArithmeticException(
                    "Factorial undefined for negative numbers"
            );
        }

        long result = 1;

        for (int i = 2; i <= n; i++) {
            result *= i;
        }

        return result;

    }

    // ==========================
    // Double Factorial
    // n!!
    // ==========================

    public static long doubleFactorial(int n) {

        if (n < 0) {
            throw new ArithmeticException(
                    "Double factorial undefined for negative numbers"
            );
        }

        long result = 1;

        for (int i = n; i > 1; i -= 2) {
            result *= i;
        }

        return result;

    }

    // ==========================
    // Super Factorial
    // sf(n) = 1! × 2! × ... × n!
    // ==========================

    public static double superFactorial(int n) {

        if (n < 0) {
            throw new ArithmeticException(
                    "Super factorial undefined for negative numbers"
            );
        }

        double result = 1.0;

        for (int i = 1; i <= n; i++) {
            result *= factorial(i);
        }

        return result;

    }

    // ==========================
    // Hyper Factorial
    // H(n) = ∏ k^k
    // ==========================

    public static double hyperFactorial(int n) {

        if (n < 0) {
            throw new ArithmeticException(
                    "Hyper factorial undefined for negative numbers"
            );
        }

        double result = 1.0;

        for (int i = 1; i <= n; i++) {
            result *= Math.pow(i, i);
        }

        return result;

    }

    // ==========================
    // Primorial
    // Product of all primes <= n
    // ==========================

    public static long primorial(int n) {

        if (n < 2) {
            return 1;
        }

        long result = 1;

        for (int i = 2; i <= n; i++) {

            if (isPrime(i)) {
                result *= i;
            }

        }

        return result;

    }

    // ==========================
    // Subfactorial
    // !n (Derangements)
    // ==========================

    public static long subFactorial(int n) {

        if (n < 0) {
            throw new ArithmeticException(
                    "Subfactorial undefined for negative numbers"
            );
        }

        if (n == 0) return 1;
        if (n == 1) return 0;

        long a = 1;
        long b = 0;

        for (int i = 2; i <= n; i++) {

            long temp = (i - 1) * (a + b);

            a = b;
            b = temp;

        }

        return b;

    }

    // ==========================
    // Prime Test
    // ==========================

    private static boolean isPrime(int n) {

        if (n < 2) {
            return false;
        }

        for (int i = 2; i * i <= n; i++) {

            if (n % i == 0) {
                return false;
            }

        }

        return true;

    }

}

File: Python/ScientificCalculator/functions/factorial.py
"""
factorial.py

Factorial functions.
"""

import math


def factorial(x):

    if int(x) != x:
        raise ValueError(
            "Factorial is only defined for integers."
        )

    return math.factorial(int(x))


def doublefactorial(x):
    """
    n!!

    Example:
    8!! = 8×6×4×2
    """

    n = int(x)

    if n < 0:
        raise ValueError("Negative numbers not allowed.")

    if n == 0 or n == 1:
        return 1

    result = 1

    while n > 1:
        result *= n
        n -= 2

    return result


def superfactorial(x):
    """
    sf(n)=1!×2!×...×n!
    """

    n = int(x)

    result = 1

    for i in range(1, n + 1):
        result *= math.factorial(i)

    return result


def hyperfactorial(x):
    """
    H(n)=1^1×2^2×...×n^n
    """

    n = int(x)

    result = 1

    for i in range(1, n + 1):
        result *= i ** i

    return result


def primefactorial(x):
    """
    Product of all primes ≤ n
    """

    n = int(x)

    result = 1

    for i in range(2, n + 1):

        prime = True

        for j in range(2, int(i ** 0.5) + 1):

            if i % j == 0:
                prime = False
                break

        if prime:
            result *= i

    return result


def risingfactorial(x, k):
    """
    (x)^k = x(x+1)...(x+k-1)
    """

    x = int(x)
    k = int(k)

    result = 1

    for i in range(k):
        result *= x + i

    return result


def fallingfactorial(x, k):
    """
    x_(k)=x(x-1)...(x-k+1)
    """

    x = int(x)
    k = int(k)

    result = 1

    for i in range(k):
        result *= x - i

    return result
